lib/iris/experimental/ugrid/__init__.py

```diff
@@ -2746,6 +2747,239 @@ def face_node(self):
+        latitude: MeshCoord
+        longitude: MeshCoord
+
+    """
+    # Validate coord essentials.
+    coord_names = {coord_1: "coord_1", coord_2: "coord_2"}
+    for coord, name in coord_names.items():
+        if not isinstance(coord, Coord):
+            message = f"Expected a Coord for {name}, got: {type(coord)}"
+            raise ValueError(message)
+
+        if coord.ndim != 1:
+            message = f"Coordinates must have ndim == 1. f{name}.ndim == {coord.ndim} ."
+            raise ValueError(message)
+
+        invalid_bounds = False
+        sub_message = ""
+        if not coord.has_bounds():
+            invalid_bounds = True
+            sub_message = "None"
+        else:
+            bounds_shape = coord.core_bounds().shape
+            # 1D coord check guarantees bounds are ndim==2.
+            if bounds_shape[1] < 2:
+                invalid_bounds = 2
+                sub_message = f"shape {bounds_shape}. Must be (n, >=2)."
+        if invalid_bounds:
+            message = f"{name} has invalid bounds: " + sub_message
+            raise ValueError(message)
+
+    # Validate orthogonality.
+    coord_axes = {
+        coord: guess_coord_axis(coord) for coord in (coord_1, coord_2)
+    }
+    if coord_axes[coord_1] is not None and coord_axes[coord_2] is not None:
+        orthogonal = coord_axes[coord_2] != coord_axes[coord_1]
+        sub_message = f"Both axes == {coord_axes[coord_1]} ."
+    elif (
+        coord_1.standard_name is not None and coord_2.standard_name is not None
+    ):
+        orthogonal = coord_2.standard_name != coord_1.standard_name
+        sub_message = f"Both standard_names == {coord_1.standard_name} ."
+    else:
+        orthogonal = coord_2.metadata != coord_1.metadata
+        sub_message = "Both coords' metadata identical."
+
+    if not orthogonal:
+        message = f"coord_2 must be orthogonal to coord_1. {sub_message}"
+        raise ValueError(message)
+
+    # Validate points and bounds shape match.
+    def check_shape(array_name):
+        attr_name = f"core_{array_name}"
+        shape_1 = getattr(coord_1, attr_name)().shape
+        shape_2 = getattr(coord_2, attr_name)().shape
+        if shape_2 != shape_1:
+            message = f"coord_2 {array_name} shape must be: {shape_1}, got: {shape_2}."
+            raise ValueError(message)
+
+    for array in ("points", "bounds"):
+        check_shape(array)
+
+    # Determine dimensionality, using coord_1.
+    bounds_shape = coord_1.core_bounds().shape
+    if bounds_shape[1] == 2:
+        topology_dimension = 1
+        coord_centring = "edge"
+        conn_cf_role = "edge_node_connectivity"
+    else:
+        topology_dimension = 2
+        coord_centring = "face"
+        conn_cf_role = "face_node_connectivity"
+
+    # Create connectivity.
+    if coord_1.has_lazy_bounds():
+        array_lib = da
+    else:
+        array_lib = np
+    indices = array_lib.arange(np.prod(bounds_shape)).reshape(bounds_shape)
```

Comment:
When the bounds array is masked this seems like it would cause the existence of nodes without location coordinates. I can't recall if we're directly supporting mixed size cells, but if we are, it's probably worth at least raising an appropriate error in this case.

Reply:
After some thought, I think we **_can_** support this quite happily, and without much further difficulty.

The key point is that building the indices like this is treating every corner of every face as an independent node
(to avoid the cost of searching the bounds arrays for duplicate values, to work out where an original node is re-visited)

In that case, I think there are just 2 problems

(1) in this connectivity array, we need to replace a 'normal' index value with a masked point,  wherever the bounds values are masked. 
I think that can just be something like...
`indices = array_lib.ma.masked_array(indices, array_lib.ma.getmaskarray(coord_1.core_bounds()))`
(i.e. apply the bounds mask to the connectivity array, but lazily if need be)

(2) in the node arrays, we will have some values which derived from masked points in the bounds array.  But we don't support masked data in these, so we should at least fill those points with something safe.
Which again is pretty easy, just use an `array_lib.ma.filled(coord.core_bounds(), 0.0)`

I think it helps to consider what a simple mixed-shape mesh looks like in the original, and then what happens when you make meshcoords for it,  and then re-construct a mesh from the meshcoord bounds.
Here is my teeny example :
```
Suppose we have a square + a triangle joined like this ...
 1--2
 |  |\
 |  | \
 3--4--5

this translates as the following
    node_x = [x1, x2, x3, x4, x5]
    face_node_connectivity = [[1, 2, 3, 4], [2, 5, 4, --]]
    face_x = [fx1, fx2]
    
(NOTE : skipping all Y values -- they behave just like the X ones )

So, if you now create MeshCoords from this, you get (X only) ...
    mesh_x.points = [fx1, fx2]
    mesh_x.bounds = [[x1, x2, x3, x4], [x2, x5, x4, --]]

If we now reconstruct a new mesh from this, it leads to the following... 
USING CAPS to denote the NEW mesh data, 
    FACE_X = [fx1, fx2]
    NODE_X = [X1=x1, X2=x2, X3=x3, X4=x4, X5=x2, X6=x5, X7=x4, ***X8=0.0*** ]  # 0.0 at "masked node"
    FACE_NODE_CONNECTIVITY = [[1, 2, 3, 4], [5, 6, 7, ***--*** ]]   # "--"==missing for unused slot
```
N.B. in this, the thing we *really* need to fix is the masked point in the connectivity array : we can't use a real index pointing to a 'masked node', as we need the missing points encoded in the connectivity itself.

Reply:
P.S. examples like the above we can calculate for real : 
We can use an indexing something like `cube = cube[..., 0:]` to remove the mesh + replace the mesh-coords with auxcoords.

Reply:
>P.S. examples like the above we can calculate for real : 
>We can use an indexing something like `cube = cube[..., 0:]` to remove the mesh + replace the mesh-coords with auxcoords.

Yes, I did something like that in the doctest:

https://github.com/SciTools/iris/blob/764e44f8883cef55ef1109ee1041700fd7c8b1c5/lib/iris/experimental/ugrid/__init__.py#L2819-L2821

Reply:
13f98539462a612dda5236c0eccdbb0bb8de02a1